### PR TITLE
[checkpointio]support asyncio for all models

### PR DESCRIPTION
**Pull request by flybird11111**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Review by wangbluo — APPROVED** [review]

**Review by wangbluo — APPROVED** [review]

**Comment by flybird11111** [comment]
checkpoint benchmark
model size: 7b gpus numbers: 4; tp=2 pp=2
use asyncio: True
image
image
use asyncio: False
image

File: colossalai/booster/plugin/torch_fsdp_plugin.py [hunk 1]
```diff
@@ -49,7 +51,10 @@ def load_unsharded_model(self, model: ModelWrapper, checkpoint: str, strict: boo
 
     def load_unsharded_optimizer(self, optimizer: OptimizerWrapper, checkpoint: Path):
         assert isinstance(optimizer, FSDPOptimizerWrapper), "Please boost the optimizer before loading!"
-        checkpoint = utils.load_state_dict(checkpoint)
+        if checkpoint.endswith(".safetensors"):
+            checkpoint = load_flat(checkpoint, seperator="-")
```
Comment (ver217) [inline colossalai/booster/plugin/torch_fsdp_plugin.py:55]: don't change seperator

File: colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py [hunk 2]
```diff
@@ -448,26 +490,46 @@ def save_sharded_optimizer(
 
         # Then collect the sharded states along dp_group(if using zero)/tp_group.
         # Only devices with (dp_rank == 0 and tp_rank == 0) are responsible for states saving.
+        control_saving = self.dp_rank == 0 and self.tp_rank == 0
```
Comment (ver217) [inline colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py:493]: should consider sp now?

File: colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py [hunk 3]
```diff
@@ -499,17 +561,30 @@ def save_sharded_optimizer(
 
             # Manage filenames of sharded weights and index file for each pipeline stage.
             states_name = states_name.replace(".bin", f"-stage-{self.pp_rank+1:05d}-shard.bin")
+            states_name = states_name.replace(".safetensors", f"-stage-{self.pp_rank+1:05d}-shard.safetensors")
```
Comment (ver217) [inline colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py:564]: These two won't occur at the same time

File: colossalai/utils/safetensors.py [hunk 4]
```diff
@@ -59,7 +59,7 @@ def _cast_to_object(tensor: torch.Tensor):
     return _tensor_to_object(tensor, tensor.numel() * tensor.element_size())
 
 
-def _flatten_optim_state_dict(state_dict: dict, seperator: str = ".") -> Tuple[dict, Optional[dict]]:
+def _flatten_optim_state_dict(state_dict: dict, seperator: str = "-") -> Tuple[dict, Optional[dict]]:
```
Comment (ver217) [inline colossalai/utils/safetensors.py:62]: keep seperator="." to keep consistent with unflatten function

File: colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py (hunk 2 above)
Comment (flybird11111) [reply colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py]: updated.

File: colossalai/utils/safetensors.py (hunk 4 above)
Comment (flybird11111) [reply colossalai/utils/safetensors.py]: updated.

File: colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py (hunk 3 above)
Comment (flybird11111) [reply colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py]: updated.

File: colossalai/booster/plugin/torch_fsdp_plugin.py (hunk 1 above)
Comment (flybird11111) [reply colossalai/booster/plugin/torch_fsdp_plugin.py]: updated.

File: colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py [hunk 5]
```diff
@@ -216,15 +244,31 @@ def save_sharded_model(
 
         # Then collect the sharded parameters & buffers along tp_group.
         # Only devices with tp_rank == 0 are responsible for model saving.
-        state_dict_shard = HybridParallelCheckpointIO._model_sharder(model, size_per_shard=size_per_shard)
+        control_saving = self.tp_rank == 0
```
Comment (ver217) [inline colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py:247]: and sp_rank == 0?

File: colossalai/utils/safetensors.py [hunk 6]
```diff
@@ -104,7 +107,11 @@ def _unflatten_optim_state_dict(flat_dict: dict, metadata: Optional[dict] = None
     for k, v in flat_dict.items():
         parts = k.split(seperator)
         assert len(parts) == 3 and parts[0] == "state"
-        idx = int(parts[1])
+        try:
+            idx = int(parts[1])
+        except:
+            # exception for fsdp, part[1] isn't param_id
+            idx = parts[1]
```
Comment (ver217) [inline colossalai/utils/safetensors.py:114]: Revert this now?

File: tests/test_checkpoint_io/test_torch_ddp_checkpoint_io.py [hunk 7]
```diff
@@ -12,14 +12,15 @@
...
 @parameterize("size_per_shard", [16, 128])
-def check_torch_ddp_checkpointIO(shard: bool, size_per_shard: int):
+@parameterize("use_async", [False, True])
+def check_torch_ddp_checkpointIO(shard: bool, size_per_shard: int, use_async: bool):
     plugin = TorchDDPPlugin()
     booster = Booster(plugin=plugin)
     model = resnet18()
     criterion = lambda x: x.mean()
-    optimizer = SGD((model.parameters()), lr=0.001)
+    optimizer = SGD((model.parameters()), lr=0.001, momentum=0.5)
```
Comment (ver217) [inline tests/test_checkpoint_io/test_torch_ddp_checkpoint_io.py:23]: keep momentum=0 to test corner case

**Review by ver217 — APPROVED** [review]
